HIRS_AttestationCA/src/main/java/hirs/attestationca/AbstractAttestationCertificateAuthority.java

```diff
@@ -1366,21 +1379,29 @@ private TPM2ProvisionerState getTpm2ProvisionerState(
 
     /**
      * Helper method to parse an Endorsement Credential from a Protobuf generated
-     * IdentityClaim. Persists the Endorsement Credential if it does not already exist.
+     * IdentityClaim. Will also check if the Endorsement Credential was already uploaded.
+     * Persists the Endorsement Credential if it does not already exist.
      *
      * @param identityClaim a Protobuf generated Identity Claim object
+     * @param ekPub the endorsement public key from the Identity Claim object
      * @return the Endorsement Credential, if one exists, null otherwise
      */
     private EndorsementCredential parseEcFromIdentityClaim(
-            final ProvisionerTpm2.IdentityClaim identityClaim) {
+            final ProvisionerTpm2.IdentityClaim identityClaim,
+            final PublicKey ekPub) {
+        EndorsementCredential endorsementCredential = null;
         if (identityClaim.hasEndorsementCredential()) {
-            return CredentialManagementHelper.storeEndorsementCredential(
+            endorsementCredential = CredentialManagementHelper.storeEndorsementCredential(
                     this.certificateManager,
                     identityClaim.getEndorsementCredential().toByteArray());
+        } else if (ekPub != null) {
+            LOG.warn("Endorement Cred was not in the identity claim from the client."
+                    + " Checking for uploads.");
+            endorsementCredential = getEndorsementCredential(ekPub);
```

Comment:
What happens if no pubek was sent?  Should that be checked for and logged?

Reply:
I'll answer your one question since I'm pretty sure they're the same. Namely, that can't happen in the current form. The IdentityClaim object is required to have an EkPublicArea so there has to be some amount of bytes there. The PublicKey itself is parsed before this gets called and if it was malformed an error would be thrown before we get to this point.